library/core/src/gc.rs

```diff
@@ -2,41 +2,51 @@
 #![allow(missing_docs)]
 
 #[cfg(not(bootstrap))]
-use crate::intrinsics;
-#[cfg(not(bootstrap))]
-use crate::mem::size_of;
-
-#[cfg(not(bootstrap))]
-static MAX_LAYOUT: usize = size_of::<usize>() * 64;
+static MAX_LAYOUT: usize = crate::mem::size_of::<usize>() * 64;
 
 #[unstable(feature = "gc", issue = "none")]
-#[cfg_attr(not(bootstrap), lang = "manageable_contents")]
-/// This trait can be implemented on types where it is safe to allow the allow the collector to
-/// free its memory and omit the drop method. This prevents the need to register a finalizer when
-/// managed by the Gc which is expensive.
-pub trait ManageableContents {}
+/// A type that implements this trait will be conservatively marked by the
+/// collector. This takes precedence over `NoTrace`.
+pub trait Conservative {}
 
 #[unstable(feature = "gc", issue = "none")]
 #[cfg_attr(not(bootstrap), lang = "no_finalize")]
 pub trait NoFinalize {}
 
+#[unstable(feature = "gc", issue = "none")]
+#[cfg_attr(not(bootstrap), lang = "notrace")]
+pub auto trait NoTrace {}
+
+#[unstable(feature = "gc", issue = "none")]
+#[derive(Debug)]
+pub struct Trace {
+    pub bitmap: u64,
+    pub size: u64,
+}
+
+#[unstable(feature = "gc", issue = "none")]
+impl Trace {
+    #[inline]
+    /// Returns true if rustgc wasn't able to create a precise descriptor for
+    /// the type.
+    pub fn is_conservative(&self) -> bool {
```

Comment:
I think maybe this function should really be called something like `fallback_to_conservative` (or something snappier) -- it's not exactly that it *is* conservative, but that we failed to make a precise tracer for this function?

Reply:
What about `must_use_conservative`? It's not much snappier but I think it's a bit more descriptive.

Reply:
Works for me!

Reply:
Fixed: 7167a888e14c0b668ffc147d18151c628ef9560e